src/lib.rs

```diff
@@ -192,6 +201,12 @@ pub mod pin {
                 bindings::analogRead(self.number()) as u16
             }
         }
+
+        pub fn register_isr(&self, edge: Edge, f: Option<extern "C" fn()>) {
```

Comment:
I don't _really_ like requiring `extern "C"` functions in a Rust API. I don't remember off the top of my head what was necessary to create a nicer API for this particular situation, though. I'm a bit surprised that there is no user data involved, because that is usually used for sending a callback. Do you know if there are any specific requirements for it? Like, does it have to be re-entrant or something? In that case we may have to mark it as `unsafe`.

Reply:
I'm pretty new to rust, so I may be wrong on this one, but as far as I understood from [https://doc.rust-lang.org/book/first-edition/ffi.html#callbacks-from-c-code-to-rust-functions](the ffi chapter of the book). 

I was also surprised about the lack of userdata, and found it super constraining, so I double checked the C library, and there really is not way to specify it. I wrote Gordon, the author of wiringpi, but the response was quite unfriendly and aggressive in a way that was totally uncalled for. Anyways the end result was that one should implement one callback function per interrupt, and then should know what caused it, and "then have that function call your mega complex giant bloaty dispatch function with the pin number". Which is a load of bullcrap IMHO, especially if one tries to dynamically assign pins based on a config file, but hey, there's nothing I could do about it. I offered to submit a PR for it, but he said fair and sqaure that he will not accept anything of the sort.

As for reentrance, it doesn't need to be, as far as i've seen the C library spawns a thread that does a poll/select on the device waiting for the interrupt, and then calls the registered function (no args).

Reply:
(So it won't be called again before the execution of the callback finishes)

Reply:
> then have that function call your mega complex giant bloaty dispatch function with the pin number

Ouch, he could at least have spared you the judgemental tone! :grimacing: He has as little business caring about anyone's bloat as we have! :triumph: Anyway, thanks for doing the extra legwork on that! I suppose it's all up to us then, and it seems like there is no unsafety involved, because we can only send function pointers to it. I'll look closer at it tomorrow (I just got annoyed by his response and had to get it off my mind before going to sleep), but I see two options as it is now; leave it as it is (not pretty, and hands the problem over to the user, but works and is fast), or build our own userdata system (for example something with `lazy_static`). I'm not a fan of quick fixes that shuffles the problem somewhere else, so I'm leaning towards the second option if it's possible.

That second option would be a global storage that associates `FnOnce() + Send + 'static` items with pin numbers, and some magic that somehow makes it all work. I don't know if it's feasible, because it doesn't even pass the pin number back... I don't know if there's some cool function pointer trick we can use to make it work.

There may also be more options, but I can't think of any right now. I'll have to save it for tomorrow.

Reply:
Uh, sorry for that, I guess you're on European time as well then? Believe me the same thing happened to me a few days before... I was just staring for a few minutes at the monitor and trying to process what I am seeing, then I got pissed and wrote him back a piece of my mind, he didn't even bother to respond.

Anyhoo I'm not sure we can do anything about it, becuase not even Fn() won't do, we actually need them to be fn() to be able to pass them to C, and we can't (easily) generate those dynamically from code. Even if we could get away with Fn() (I saw an unsupported note somewhere that we could get away with it IF the closure but only if it doesn't capture any values, which we can't guarantees work when coming from user code), we're back to square one, we don't have any data.

Also since this is mostly a wrapper around wiringpi, my position is to expose this function as it is implemented by wiringpi, and if we can find a solution, we could offer an alternative, higher level implementation under a different name.

Btw, in my code I wound up using a `const mut` to save the actor address and message that needs to be notified (I'm using actix as a bloatware library :D ).

Reply:
> Uh, sorry for that, I guess you're on European time as well then?

Yep, but don't worry about it. I'm used to being out of sync. :smile: And you are not to blame for anything, as far as I'm concerned. That kind of response is unnecessary, even if provoked, in my opinion. But let's not linger on that.

> Also since this is mostly a wrapper around wiringpi, my position is to expose this function as it is implemented by wiringpi, and if we can find a solution, we could offer an alternative, higher level implementation under a different name.

That's a good point. I'm only passively keeping this library alive as long as people are using it, so let's not complicate it more than necessary. I'm hoping that it will become boring enough for someone to replace it. I'm grateful for the help, though!

You are right, there is no nice way to do it without having a finite number of functions in some kind of pool, as far as I can tell. That would need to be updated in the future if a device with more pins shows up. It's less restrictive to just leave it as it is and let the user decide what to do.

Would you mind adding a documentation comment that briefly explains that the C function is required by wiringpi? Maybe also a simple example, if possible. Doesn't have to be more fancy than just printing something in the callback. I think that should be all that's left.

> Btw, in my code I wound up using a `const mut` to save the actor address and message that needs to be notified (I'm using actix as a bloatware library :D ).

I would recommend looking into https://crates.io/crates/once_cell for a nice safety layer instead of `const mut`. :slightly_smiling_face: It's essentially the same thing, but prevents some possible mistakes.

Reply:
unhide this conv